### ISSUE TITLE
SD scripts flux.dev 训练配置文件

### ISSUE THREAD

**Issue by twn39** [twn39/code#434]
配置文件：

```toml
output_dir = "/root/autodl-tmp/flux/model"
output_name = "simon_bailly_flux"
train_data_dir = "/root/train_data"
logging_dir = "/root/tf-logs"

pretrained_model_name_or_path = "/root/autodl-tmp/flux1-dev.safetensors"
ae = "/root/autodl-tmp/ae.safetensors"
clip_l = "/root/autodl-tmp/clip_l.safetensors"
t5xxl = "/root/autodl-tmp/t5xxl_fp16.safetensors"

max_train_steps = 4600
train_batch_size = 1
epoch = 10
unet_lr = 0.0001
resolution = "1024,1024"
discrete_flow_shift = 3.1582
optimizer_args = ["weight_decay=0.01","eps=1e-08","betas=(0.9, 0.999)"]
optimizer_type = "AdamW8bit"
network_dim = 32
network_alpha = 16

bucket_no_upscale = true
bucket_reso_steps = 64
huber_c = 0.1
huber_scale = 1
huber_schedule = "snr"
cache_latents = true
cache_latents_to_disk = true
cache_text_encoder_outputs = true
cache_text_encoder_outputs_to_disk = true
caption_extension = ".txt"
dynamo_backend = "no"
enable_bucket = true
fp8_base = true
gradient_checkpointing = true
guidance_scale = 1.0
highvram = true
loss_type = "l2"
lr_scheduler = "cosine_with_restarts"
lr_scheduler_args = []
lr_scheduler_num_cycles = 1
lr_scheduler_power = 1
lr_warmup_steps = 0
max_bucket_reso = 2048
max_data_loader_n_workers = 2
max_grad_norm = 1
max_timestep = 1000
min_bucket_reso = 256
min_snr_gamma = 5
mixed_precision = "bf16"
model_prediction_type = "raw"
network_args = [ "train_double_block_indices=all", "train_single_block_indices=all",]
network_module = "networks.lora_flux"
network_train_unet_only = true
persistent_data_loader_workers = 1
save_every_n_epochs = 1
save_model_as = "safetensors"
save_precision = "bf16"
sdpa = true
seed = 42
t5xxl_max_token_length = 512
timestep_sampling = "sigmoid"
sigmoid_scale = 1.0
noise_offset = 0.03
```

训练命令：

```bash
accelerate launch --dynamo_backend no --dynamo_mode default --mixed_precision bf16 --num_processes 1 --num_machines 1 --num_cpu_threads_per_process 2 /root/sd-scripts/flux_train_network.py --config_file /root/config_lora.toml
```

数据集文件命名：`repeats_class instance`, 例如：`10_blindbox blindbox`

其它具体参数参考：https://github.com/kohya-ss/sd-scripts/tree/sd3?tab=readme-ov-file#flux1-training

**Comment by twn39** [comment]
# SDXL

```config_sdxl_lora.toml
# ------------------------

output_name = "Jang_Wonyoung_xl_v2"
train_data_dir = "/root/zyy"
output_dir = "/root/autodl-tmp/loras/model"
pretrained_model_name_or_path = "/root/autodl-tmp/epicrealismXL_vx1Finalkiss.safetensors"
logging_dir = "/root/tf-logs"
train_batch_size = 1

unet_lr = 1e-4           # 学习率
resolution = "1024,1024"  # 训练分辨率   
optimizer_args = ["weight_decay=0.01","eps=1e-08","betas=(0.9, 0.999)"] # 优化器参数
optimizer_type = "AdamW8bit"  # Dadaptation, AdamW8bit, AdamWScheduleFree, Adafactor
lr_scheduler = "cosine_with_restarts" # constant, constant_with_warmup, cosine, cosine_with_restarts
max_train_steps = 3500  # 最大训练步数： repeat_num * images_num * epoch / batch_size
network_dim = 128
network_alpha = 64
epoch = 10

# ---------------------------

bucket_no_upscale = true
bucket_reso_steps = 64
cache_latents = true
caption_extension = ".txt"
dynamo_backend = "no"
enable_bucket = true
gradient_accumulation_steps = 1
gradient_checkpointing = true
highvram = true
huber_c = 0.1
huber_scale = 1
huber_schedule = "snr"
# loraplus_lr_ratio = 16
# loraplus_text_encoder_lr_ratio = 16
loraplus_unet_lr_ratio = 16
loss_type = "l2"
lr_scheduler_args = []
lr_scheduler_num_cycles = 2
lr_scheduler_power = 1
lr_warmup_steps = 0.1
max_bucket_reso = 2048
max_data_loader_n_workers = 2
max_grad_norm = 1
max_timestep = 1000
max_token_length = 150
min_bucket_reso = 256
min_snr_gamma = 5
persistent_data_loader_workers = 1
mixed_precision = "fp16"
network_args = []
network_module = "networks.lora"
network_train_unet_only = true
no_half_vae = true            # 开启避免 sdxl vae 黑屏
noise_offset = 0.0375
noise_offset_type = "Original"
prior_loss_weight = 1
save_every_n_epochs = 1
save_model_as = "safetensors"
save_precision = "fp16"
sdpa = true
seed = 42
text_encoder_lr = []
```


```
accelerate launch --dynamo_backend no --dynamo_mode default --mixed_precision fp16 --num_processes 1 --num_machines 1 --num_cpu_threads_per_process 2 /root/sd-scripts/sdxl_train_network.py --config_file /root/config.toml   
```


蓝色曲线配置： 

- steps: 4600
- lr: 1e-4
- restart_cycles: 2
- epoch: 20
- repeat number: 5
- batch size: 1

紫色曲线配置： 

- steps: 4600
- lr: 1e-4
- restart_cycles: 2
- epoch: 10
- repeat number: 10
- batch size: 1

绿色曲线配置： 

- steps: 4600
- lr: 1e-4
- restart cycles: 1
- epoch: 10
- repeat number: 10
- batch size: 1


<img width="2162" height="990" alt="Image" src="https://github.com/user-attachments/assets/a2db6527-6379-4be6-b779-2b13ce1accf8" />
<img width="2170" height="1002" alt="Image" src="https://github.com/user-attachments/assets/57a4750e-df78-45e4-a988-5996bd9a0dfc" />

**总结：**

`epoch: 10, repeat number: 10` 学习率波动更小，所以并非 epoch 越大越好，`restart cycles` 参数 1 和 2，在第一个周期学习率重合 ，在第二个周期参数 1 比参数 2 loss 会更低，但是总体趋势一致，三条曲线都可以得出最佳权重为第 7 个 epoch，得出最佳设置：

```
- steps: 4600
- lr: 1e-4
- restart cycles: 1
- epoch: 10
- repeat number: 10
- batch size: 1
```

**Comment by twn39** [comment]
```tags.py
import PIL.Image
import os
from openai import OpenAI
import base64
import io


directory_path = r"C:\Users\Administrator\Desktop\zyy\result"
api_key='sk-or-v1-xxx'
trigger_words = r"zyy, photography, realistic"

client = OpenAI(
  base_url="https://openrouter.ai/api/v1",
  api_key=api_key,
)

def resize_image_if_needed(img):
    width, height = img.size
    # 判断长边是否大于1024
    if max(width, height) > 1024:
        if width >= height:
            new_width = 1024
            new_height = int(height * 1024 / width)
        else:
            new_height = 1024
            new_width = int(width * 1024 / height)
        img = img.resize((new_width, new_height), PIL.Image.Resampling.LANCZOS)
    return img

def encode_image_from_pil(img):
    buffered = io.BytesIO()
    # 使用JPEG保存图片，如果原图片支持透明通道则保存PNG
    img_format = 'PNG' if img.mode in ('RGBA', 'LA') else 'JPEG'
    img.save(buffered, format=img_format)
    return base64.b64encode(buffered.getvalue()).decode('utf-8')


def save_string_to_txt(string, image_path):
    # 获取图片文件名（不包含扩展名）
    file_name = os.path.splitext(os.path.basename(image_path))[0]
    # 构建文本文件路径
    txt_file_path = os.path.join(directory_path, f"{file_name}.txt")
    print(txt_file_path)
    # 保存字符串到文本文件
    with open(txt_file_path, 'w', encoding="utf-8") as file:
        file.write(string)


image_files = [filename for filename in os.listdir(directory_path)
               if filename.lower().endswith((".jpg", ".jpeg", ".png", ".webp"))]
image_files.sort()

prompt = '''
This is a photo of Korean young woman. Please use about 30 words in English to describe the contents of the photo, including the clothing, hairstyle, and action.
 Do not describe facial details, and do not add explanatory words like: 'Here's a description of the photo'. Describe the content directly. 
 Example: 
 a young woman with long, dark hair and bangs, wearing a white sleeveless top, poses with one hand making a peace sign near her face, in an office setting.
'''

for filename in image_files:
    filepath = os.path.join(directory_path, filename)
    print(filepath)
    img = PIL.Image.open(filepath)
    img = resize_image_if_needed(img)
    base64_image = encode_image_from_pil(img)
    completion = client.chat.completions.create(
        model="google/gemini-2.0-pro-exp-02-05:free",
        temperature=0.1,
        messages=[{
            "role": "user",
            "content": [
                {"type": "text", "text": prompt},
                {"type": "image_url", "image_url": {
                    "url": f"data:image/jpeg;base64,{base64_image}"
                },},
            ],
        }],
    )
    result = completion.choices[0].message.content
    print(result)
    save_string_to_txt(f"{trigger_words}, {result}", filepath)

```
